src/library/scala/collection/BitSet.scala

```diff
@@ -108,15 +108,23 @@ trait BitSetOps[+C <: BitSet with BitSetOps[C]]
   def iterator: Iterator[Int] = iteratorFrom(0)
 
   def iteratorFrom(start: Int): Iterator[Int] = new AbstractIterator[Int] {
-    private[this] var current = start
-    private[this] val end = nwords * WordLength
-    def hasNext: Boolean = {
-      while (current != end && !self.contains(current)) current += 1
-      current != end
+    private[this] var currentPos = if (start > 0) start >> LogWL else 0
+    private[this] var currentWord = if (start > 0) word(currentPos) & (-1L << (start & (WordLength - 1))) else word(0)
+    final override def hasNext: Boolean = {
+      while (currentWord == 0) {
+        if (currentPos + 1 >= nwords) return false
```

Comment:
@linasm I'm not a fan of `return` in Scala as it breaks last-expr-is-the-result assumptions.

Reply:
@viktorklang me neither, but I feel similar about `tailrec` method that does side effects :)
Before submitting this change, I saw `return` from `while` all over `BitSet` implementation:
https://github.com/scala/scala/blob/4ec48ff29755fc01fdde880dbb4548085420c601/src/library/scala/collection/BitSet.scala#L148-L151
https://github.com/scala/scala/blob/4ec48ff29755fc01fdde880dbb4548085420c601/src/library/scala/collection/BitSet.scala#L160-L163
https://github.com/scala/scala/blob/4ec48ff29755fc01fdde880dbb4548085420c601/src/library/scala/collection/mutable/BitSet.scala#L246-L247
https://github.com/scala/scala/blob/4ec48ff29755fc01fdde880dbb4548085420c601/src/library/scala/collection/mutable/BitSet.scala#L253-L254
Not implying that this legitimizes the pattern or something, just would appreciate a second opinion from someone who maintains this part of the code base.

Reply:
@linasm I think "prior art" is a valid argument. :)

Reply:
IMHO, while "prior art" is a fair enough reason, there is no reason not to "clean" it along the way, unless it defeats performance of course. I think the following should work (but please do test first).

```Scala
      while (currentWord == 0 && currentPos + 1 < nwords) {
        currentPos += 1
        currentWord = word(currentPos)
      }
      currentWord != 0 || currentPos + 1 < nwords
```

Reply:
@diesalbla your implementation seems to be correct, but it is slightly slower in most cases:
```
[info] Benchmark                           (spacing)  Mode  Cnt     Score    Error  Units
[info] BitSetIteratorBenchmark.iterateAll          0  avgt    6  3561.769 ± 79.572  ns/op
[info] BitSetIteratorBenchmark.iterateAll          1  avgt    6  1786.156 ± 11.039  ns/op
[info] BitSetIteratorBenchmark.iterateAll          3  avgt    6   934.819 ± 13.076  ns/op
[info] BitSetIteratorBenchmark.iterateAll         15  avgt    6   253.551 ±  5.492  ns/op
[info] BitSetIteratorBenchmark.iterateAll         63  avgt    6    83.190 ±  0.820  ns/op
[info] BitSetIteratorBenchmark.iterateAll        255  avgt    6    38.085 ±  3.758  ns/op
```
`hasNext` is normally invoked twice for each advancement of the iterator (once directly from client code, and once from `next()`), and in most invocations it does not enter the `while` loop. I've optimized my code under this assumption, making sure that just one comparison is done in those cases.